### PR TITLE
RavenDB-5039 First stab at virtual grid in edit document page.

### PR DESCRIPTION
**Pull request by JudahGabriel**
Virtual grid now supports hiding column headers, column-based cell rendering, and hyperlink cells. We aren't finished integrating into the edit doc page: while rows show up and are clickable links, howerver, we need to implement search filtering and collection changing between related/collection/recent/starred.

### CODE REVIEW

File: src/Raven.Studio/typescript/viewmodels/database/documents/editDocumentConnectedDocuments.ts [hunk 1]
```diff
@@ -52,6 +43,46 @@ class connectedDocuments {
...
+        starred: "starred"
+    };
+    static currentTab = ko.observable<string>(connectedDocuments.connectedDocsTabs.related);
+
+    constructor(document: KnockoutObservable<document>, db: KnockoutObservable<database>, loadDocument: (docId: string) => void) {
+        this.document = document;
+        this.db = db;
+        this.document.subscribe((doc) => this.onDocumentLoaded(doc));
+        this.loadDocumentAction = loadDocument;
+        virtualGrid.install();
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/documents/editDocumentConnectedDocuments.ts:59]: Should we install this extension each time we create view?

File: src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿import virtualColumn = require("widgets/virtualGrid/virtualColumn");
+
+/**
+ * Checked cell template used for toggling selection of a row in the virtual grid.
+ */
+class checkedColumn implements virtualColumn {
+    public readonly cellClass = "checked-cell";
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:7]: public is default in typescript so there is no need to write this. (we are following the default resharper code style suggestions except class names starts with lower case)

File: src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    public readonly dataMemberName = "__virtual-grid-isChecked";
+    public readonly width = "32px";
+    public readonly display = `<input class="checked-column-header" type="checkbox" />`;
+
+    public static readonly columnWidth = 32;
+
+    constructor() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:14]: Do we need this empty constructor?

File: src/Raven.Studio/typescript/widgets/virtualGrid/hyperlinkColumn.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿import virtualColumn = require("widgets/virtualGrid/virtualColumn");
+
+/**
+ * Virtual grid column that renders hyperlinks.
+ */
+class textColumn implements virtualColumn {
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/hyperlinkColumn.ts:6]: class hyperlink Column?

File: src/Raven.Studio/typescript/widgets/virtualGrid/textColumn.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿import virtualColumn = require("widgets/virtualGrid/virtualColumn");
+
+/**
+ * Virtual grid column that renders text in its cells.
+ */
+class textColumn implements virtualColumn {
+    public readonly cellClass = "text-cell";
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/textColumn.ts:7]: remove public

File: src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts [hunk 6]
```diff
@@ -349,34 +352,35 @@ class VirtualGrid<T> {
...
-}
-
-// TODO: we may want to move this to an HTML file that's fetched with RequireJS. Knockout components do support this.
-ko.components.register("virtual-grid", {
-    viewModel: VirtualGrid,
-    template: `
+    /**
+     * Installs the virtual grid component if it's not already installed.
+     */
+    public static install() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts:358]: remove public

File: src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts [hunk 7]
```diff
@@ -349,34 +352,35 @@ class VirtualGrid<T> {
...
+        const componentName = "virtual-grid";
+        if (!ko.components.isRegistered(componentName)) {
```
Comment (ml054) [inline src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts:360]: maybe we should init this in shell? or extensions.ts class?

File: src/Raven.Studio/wwwroot/Content/css/virtual-grid.less [hunk 8]
```diff
@@ -0,0 +1,79 @@
+﻿.virtual-grid {
+	overflow-x: auto;
+	overflow-y: hidden;
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/Content/css/virtual-grid.less:3]: this file has issues with indentation?

File: src/Raven.Studio/wwwroot/Content/css/virtual-grid.less [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    .viewport-scroller {
+	    position: relative;
+    }
+
+    .column {
+	    display: inline-block;
+	    text-overflow: ellipsis;
+	    overflow: hidden;
+	    text-align: left;
+	    border-bottom: 2px solid rgb(221, 221, 221);
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/Content/css/virtual-grid.less:39]: we should avoid specifing colors directly, @kwiato will change this to variables, so we can later support color themes.

**Review by ml054 — CHANGES_REQUESTED** [review]

**Comment by ayende** [comment]
@JudahGabriel Can you go over the changes?

File: src/Raven.Studio/typescript/viewmodels/database/documents/editDocumentConnectedDocuments.ts (hunk 1 above)
Comment (JudahGabriel) [reply src/Raven.Studio/typescript/viewmodels/database/documents/editDocumentConnectedDocuments.ts:59]: That, or we can install it once (maybe in shell). It only needs to be installed once. Install will do nothing if it's already installed.

File: src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts (hunk 2 above)
Comment (JudahGabriel) [reply src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:7]: Sure, I'll remove it. p.s. I agree with Resharper we should eventually move class names to start with upper case. That seems to be the emerging standard for TypeScript anyways.

File: src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts (hunk 3 above)
Comment (JudahGabriel) [reply src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:14]: Nope. I'll remove it.

File: src/Raven.Studio/typescript/widgets/virtualGrid/hyperlinkColumn.ts (hunk 4 above)
Comment (JudahGabriel) [reply src/Raven.Studio/typescript/widgets/virtualGrid/hyperlinkColumn.ts:6]: Oh, that's...weird. I'll fix that.

File: src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts (hunk 7 above)
Comment (JudahGabriel) [reply src/Raven.Studio/typescript/widgets/virtualGrid/virtualGrid.ts:360]: Yep, that's fine. We can do a one-time initialization in shell. I'll do that.

**Comment by JudahGabriel** [comment]
OK. I will fix these issues, squash the commits into one, and issue a new PR.

File: src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts (hunk 2 above)
Comment (ml054) [reply src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:7]: I agree but it is a lot of work and we went with this convention due to pvc / windows path issues. so before changing this we have to investigate potential issues.

Comment (JudahGabriel) [reply src/Raven.Studio/typescript/widgets/virtualGrid/checkedColumn.ts:7]: Yep, understood. I suggest we move to upper case classes at the time we migrate the Studio to Angular/React/Aurelia post-4.0.

**Pull request closed without merging**